crates/medea-macro/src/lib.rs

```diff
@@ -220,3 +222,65 @@ pub fn enum_delegate(args: TokenStream, input: TokenStream) -> TokenStream {
 pub fn dispatchable(_: TokenStream, input: TokenStream) -> TokenStream {
     dispatchable::derive(input).unwrap_or_else(|e| e.to_compile_error().into())
 }
+
+decl_derive!([JsCaused, attributes(js_cause, js_error)] =>
+/// Generate implementation `JsCaused` trait for errors represented as `enum`.
+///
+/// # How to use
+///
+/// ### 1. Declare wrapper for JS error and `enum` for error variants.
+///
+/// The `js_cause()` method returns error if nested error has type declare as
+/// argument of `js_error` attribute like `#[js_error(path::to::Error)]`.
+///
+/// ```
+/// use medea_jason::utils::JsCaused;
+///
+/// struct JsError;
+///
+/// #[derive(JsCaused)]
+/// #[js_error(JsError)]
+/// enum FooError {
+///     Internal,
+///     Js(JsError),
+/// }
+///
+/// let err = FooError::Internal;
+/// assert_eq!(err.name(), "Internal");
+/// assert!(err.js_cause().is_none());
+///
+/// let err = FooError::Js(JsError {});
+/// assert_eq!(err.name(), "Js");
+/// assert!(err.js_cause().is_some());
+/// ```
+///
+/// If enum variant has attribute `#[js_caused]` it will be called `js_cause()`
+/// on nested error.
+///
+/// ```
+/// # use medea_jason::utils::JsCaused;
+/// #
+/// # struct JsError;
+/// #
+/// # #[derive(JsCaused)]
+/// # #[js_error(JsError)]
+/// # enum FooError {
+/// #     Internal,
+/// #     Js(JsError),
+/// # }
+///
+/// #[derive(JsCaused)]
+/// #[js_error(JsError)]
+/// enum BarError {
+///     Foo(#[js_cause] FooError),
```

Comment:
Also, let's bikeshed attr names a bit like the following:
```rust
#[derive(JsCaused)]
#[js(error = JsError)]
enum BarError {
    Foo(#[js(cause)] FooError),
}
```
which is more idiomatic in ecosystem (see at `derive_more` attrs, for example).